TASK: Tests for "Content Repository Privileges"

bwaidelich opened the pull request:
Related: #3732
Related: #5298

Behat tests for #3732

bwaidelich committed:
TASK: Tests for "Content Repository Privileges"

Related: #3732
Related: #5298

bwaidelich committed:
moar tests

bwaidelich committed:
moar better tests

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-tests

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/ContentRepositorySecurityTrait.php:
```diff
@@ -0,0 +1,210 @@
+
+    private ?TestingProvider $testingProvider = null;
+
+    private ?ActionRequest $mockActionRequest = null;
```
this needs to be prefixed or something to avoid clash of other traits ... this is tooo common.
```suggestion
    private ?ActionRequest $crSecurity_actionRequest = null;
```

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/ContentRepositorySecurityTrait.php:
```diff
@@ -0,0 +1,210 @@
+        $policyConfiguration = ObjectAccess::getProperty($policyService, 'policyConfiguration', true);
+        $mergedPolicyConfiguration = Arrays::arrayMergeRecursiveOverrule($policyConfiguration, Yaml::parse($policies->getRaw()));
+
+        self::$testingPolicyPathAndFilename = $this->getObject(Environment::class)->getPathToTemporaryDirectory() . 'Policy.yaml';
+        file_put_contents(self::$testingPolicyPathAndFilename, Yaml::dump($mergedPolicyConfiguration));
+
+        ObjectAccess::setProperty($policyService, 'initialized', false, true);
+        $this->getObject(ConfigurationManager::class)->flushConfigurationCache();
```
is this latter part really needed? As we dont have sub processes we dont have to preserve the state and mess with the file system right?

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/ContentRepository/Security/EditNodePrivilege.feature:
```diff
@@ -0,0 +1,74 @@
+      | nodeAggregateId              | "a1a"         |
+      | nodeVariantSelectionStrategy | "allVariants" |
+    Then the last command should have thrown an exception of type "AccessDenied" with code 1729086686
+#    Then the last command should have thrown an exception of type "AccessDenied" with message:
+#    """
+#    Command "Neos\ContentRepository\Core\Feature\NodeDisabling\Command\DisableNodeAggregate" was denied: No edit permissions for node "a1a" in workspace "live": Evaluated following 2 privilege target(s):
+#    "Neos.Neos:ReadBlog": ABSTAIN
+#    "Neos.Neos:ReadBlog": GRANT
+#    (1 granted, 0 denied, 1 abstained)
+#    Evaluated following 1 privilege target(s):
+#    "Neos.Neos:EditBlog": ABSTAIN
+#    (0 granted, 0 denied, 1 abstained)
+#    """
```
todo

bwaidelich replied:
WIP ;)

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/ContentRepository/Security/ReadNodePrivilege.feature:
```diff
@@ -0,0 +1,68 @@
+      | johndoe | john.doe | John       | Doe       | Neos.Neos:RestrictedEditor,Neos.Neos:UserManager |
+      | editor  | editor   | Edward     | Editor    | Neos.Neos:Editor                                 |
+
+  Scenario: TODO
```
;)

mhsdesign reviewed (commented):
I know this is wip and all but i wanted to see where were going :)

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

bwaidelich committed:
BUGFIX: Behat: Reset `lastCommandException` once asserted

Without this fix it was not possible to expect multiple exceptions in a single step or, worse, it led to false positives:

```gherkin
  Scenario:
    When doing something
    Then the last command should have thrown an exception of type "Foo"

    When doing something else
    Then the last command should have thrown an exception of type "Bar"
```

The 2nd time would succeed even if the previous interaction did not lead to an exception

mhsdesign committed:
TASK: Enforce command exceptions are asserted

mhsdesign committed:
TASK: Improve lastCommandException assertions

kitsunet committed:
Merge pull request #5362 from neos/bugfix/reset-lastCommandException-once-asserted-in-behat-tests

BUGFIX: Behat: Reset `lastCommandException` once asserted

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php

bwaidelich committed:
Add more tests

bwaidelich committed:
Prefix ContentRepositorySecurityTrait fields to avoid naming clashes
